controlplane/kubeadm/internal/controllers/controller_test.go

```diff
@@ -749,6 +756,150 @@ func TestKubeadmControlPlaneReconciler_adoption(t *testing.T) {
 	})
 }
 
+func TestKubeadmControlPlaneReconciler_ensureOwnerReferences(t *testing.T) {
+	g := NewWithT(t)
+
+	cluster, kcp, tmpl := createClusterWithControlPlane(metav1.NamespaceDefault)
+	cluster.Spec.ControlPlaneEndpoint.Host = "bar"
+	cluster.Spec.ControlPlaneEndpoint.Port = 6443
+	cluster.Status.InfrastructureReady = true
+	kcp.Spec.Version = "v1.21.0"
+	key, err := certs.NewPrivateKey()
+	g.Expect(err).To(BeNil())
+	crt, err := getTestCACert(key)
+	g.Expect(err).To(BeNil())
+
+	fmc := &fakeManagementCluster{
+		Machines: collections.Machines{},
+		Workload: fakeWorkloadCluster{},
+	}
+
+	clusterSecret := &corev1.Secret{
+		// The Secret's Type is used by KCP to determine whether it is user-provided.
+		// clusterv1.ClusterSecretType signals that the Secret is CAPI-provided.
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      "",
+			Labels: map[string]string{
+				"cluster.x-k8s.io/cluster-name": cluster.Name,
+				"testing":                       "yes",
+			},
+		},
+		Data: map[string][]byte{
+			secret.TLSCrtDataName: certs.EncodeCertPEM(crt),
+			secret.TLSKeyDataName: certs.EncodePrivateKeyPEM(key),
+		},
+	}
+
+	t.Run("add KCP owner for secrets with no controller reference", func(t *testing.T) {
+		objs := []client.Object{fakeGenericMachineTemplateCRD, cluster.DeepCopy(), kcp.DeepCopy(), tmpl.DeepCopy()}
+		for _, purpose := range []secret.Purpose{secret.ClusterCA, secret.FrontProxyCA, secret.ServiceAccount, secret.EtcdCA} {
+			s := clusterSecret.DeepCopy()
+			// Set the secret name to the purpose
+			s.Name = secret.Name(cluster.Name, purpose)
+			// Set the Secret Type to clusterv1.ClusterSecretType which signals this Secret was generated by CAPI.
+			s.Type = clusterv1.ClusterSecretType
+
+			objs = append(objs, s)
+		}
+
+		fakeClient := newFakeClient(objs...)
+		fmc.Reader = fakeClient
+		r := &KubeadmControlPlaneReconciler{
+			Client:                    fakeClient,
+			APIReader:                 fakeClient,
+			managementCluster:         fmc,
+			managementClusterUncached: fmc,
+		}
+
+		_, err := r.reconcile(ctx, cluster, kcp)
+		g.Expect(err).To(BeNil())
+
+		secrets := &corev1.SecretList{}
+		g.Expect(fakeClient.List(ctx, secrets, client.InNamespace(cluster.Namespace), client.MatchingLabels{"testing": "yes"})).To(Succeed())
+		for _, secret := range secrets.Items {
+			g.Expect(secret.OwnerReferences).To(ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
+		}
+	})
+
+	t.Run("replace non-KCP controller with KCP controller reference", func(t *testing.T) {
```

Comment:
Should we have cases where we preserve existing third-party non-controller refs? (maybe integrate in this case and the next one?)